### PR TITLE
Stage1 mvp batch dimension

### PR DESCRIPTION
**Pull request by ryk-wolf**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Ensure that code and tests are properly formatted, by running `make format` or `black -l 100
      <filename>` on any relevant files. You will need to have the Black code format installed:
      `pip install black`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] The Mr Mustard source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/) except line length.
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint mrmustard/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: mrmustard/lab/representations/data/array_data.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        try:
+            return self.__class__(array=self.array / other)
+        except TypeError as e:
+            raise TypeError("Can only divide by a scalar.") from e
+
+    def __mul__(self, other: Scalar) -> ArrayData:
+        try:
+            return self.__class__(array=self.array * other)
+        except TypeError as e:
+            raise TypeError("Can only multiply by a scalar.") from e
```
Comment (ziofil) [inline mrmustard/lab/representations/data/array_data.py:65]: if `other` is an array that has compatible shape with `self.array`, the multiplication will still be successful. So we aren't only allowing multiplication by scalars.

File: mrmustard/lab/representations/data/data.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+from abc import ABC, abstractmethod
+from typing import List, Union
+
+from mrmustard.typing import Scalar, Vector
+
+
+class Data(ABC):
+    r"""Abstract parent class for types of data encoding a quantum state's representation."""
+
+    def __init__(self) -> None:
```
Comment (ziofil) [inline mrmustard/lab/representations/data/data.py:28]: I don't think ABC classes should have an init method

File: mrmustard/lab/representations/data/data.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+        Returns:
+            (bool) True if all the elements compared are same within numpy's default rtol/atol, 
+            False otherwise
+        """
+        return all([np.allclose(x, y) for x, y in zip(X, Y)])
+
+    @abstractmethod
+    def __neg__(self) -> Data:
+        raise NotImplementedError()  # prompting override in children
```
Comment (ziofil) [inline mrmustard/lab/representations/data/data.py:50]: Just a style comment: this method is already marked as abstract, so it won't be possible to instantiate children of this class without implementing it. In other words, the NotImplementedError is redundant.

File: mrmustard/lab/representations/data/gaussian_data.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+        self,
+        cov: Optional[Batch[Matrix]] = None,
+        means: Optional[Batch[Vector]] = None,
+        coeffs: Optional[Batch[Scalar]] = None,
+    ) -> None:
+        
+        if cov is not None or means is not None:  # at least one is defined -or both-
+
+            if cov is None:
+                means =  np.array(means)
```
Comment (ziofil) [inline mrmustard/lab/representations/data/gaussian_data.py:62]: we shouldn't use numpy's methods, or we lose differentiability. Perhaps you're looking for `math.astensor`

File: mrmustard/lab/representations/data/gaussian_data.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+                dim = means.shape[1]
+                batch_size = means.shape[0]
+                cov = np.array([math.eye(dim, dtype=means.dtype) for _ in range(batch_size)])
+
+            elif means is None:  # we know cov is not None here
+                means =  np.array(means)
```
Comment (ziofil) [inline mrmustard/lab/representations/data/gaussian_data.py:68]: same here and line below

File: mrmustard/lab/representations/data/gaussian_data.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        r"""Computes the combined covariances when multiplying Gaussian-represented states.
+
+        Args:
+            other (GaussianData): another GaussianData object which covariance will be multiplied
+
+        Returns:
+            (Tensor) The tensor of combined covariances
+        """
+        # NOTE: will this get solved when the batch dimension is supported everywhere?
+        # c1 = math.expand_dims(c1, axis=0)
```
Comment (ziofil) [inline mrmustard/lab/representations/data/gaussian_data.py:119]: shouldn't we use `math_atleast_1d` / `2d` /  `3d` ?

**Comment by elib20** [comment]
Just switched this to draft until some issues get resolved

**Pull request closed without merging**